Feature/service status update

mked-luxoft opened the pull request:
Implements #2791

This PR is **[ready]** for review.

### Risk
This PR makes **[minor]** API changes.

### Testing Plan
Unit tests
ATF tests

### Summary
Added ServiceStatusUpdateHandler and ServiceStatusUpdateHandlerListener interface to dispatch OnServiceUpdate notifications depending on handshake status.

[ServiceStatusUpdate_classes.pdf](https://github.com/Ford-Luxoft/sdl_core/files/2791518/ServiceStatusUpdate_classes.pdf)

[ServiceStatusUpdate_sequence.pdf.pdf](https://github.com/Ford-Luxoft/sdl_core/files/2791519/ServiceStatusUpdate_sequence.pdf.pdf)




### Tasks Remaining:
- [ ] Unit tests
- [ ] ATF tests
- [ ] Bugfix
- [ ] Clean up (remove duplications, remove redundant code)

### CLA
- [ ] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

LuxoftAKutsan commented on src/appMain/life_cycle_impl.cc:
```diff
@@ -103,6 +103,14 @@ bool LifeCycleImpl::StartComponents() {
   app_manager_ =
       new application_manager::ApplicationManagerImpl(profile_, profile_);
 
+  auto service_status_update_handler =
```
@mked-luxoft what about life cycle of this object.

mked-luxoft replied:
@LuxoftAKutsan  it is set to `protocol_handler` instance and remains its member throughout its lifecycle

LuxoftAKutsan replied:
@mked-luxoft can we use unique ptr? And move it to holder?

mked-luxoft replied:
@LuxoftAKutsan Sorry, missed one point. Since `service_status_update_handler` instance is also passed to `HandshakeHandler` by `ProtocolHandler`, i think it's better to stick with `std::shared_ptr`

LuxoftAKutsan replied:
@mked-luxoft and can HandshakeHandler keep ServiceStatusUpdateHandler by listener?

mked-luxoft replied:
@LuxoftAKutsan as i understand in this case, `HandshakeHandler` will need to depend on API headers.

AKalinich-Luxoft replied:
@mked-luxoft I think you should call `reset()` for `service_status_update_handler` in StopComponents as its done for all other components

LuxoftAKutsan commented on src/components/application_manager/include/application_manager/application_manager_impl.h:
```diff
@@ -509,6 +511,19 @@ class ApplicationManagerImpl
    */
   void OnPTUFinished(const bool ptu_result) FINAL;
 
+  void ProcessFailedPTU() OVERRIDE;
```
@mked-luxoft why this is override but others are final?

LuxoftAKutsan commented on src/components/application_manager/include/application_manager/application_manager_impl.h:
```diff
@@ -509,6 +511,19 @@ class ApplicationManagerImpl
+      hmi_apis::Common_ServiceType::eType service_type,
+      hmi_apis::Common_ServiceEvent::eType service_event) FINAL;
+
+  void OnPTUFailed() FINAL {}
```
@mked-luxoft misundestanding in functions naming OnPTUFailed andProcessFailedPTU

mked-luxoft replied:
@LuxoftAKutsan  `ProcessFailedPTU` is an interface that allows notifying `protocol_handler` that PTU was unsuccessful. On the other hand, `OnPTUFailed` is a part of `SecurityManagerListener` listener interface. This callback exists, so that `HandshakeHandler`  (which implements `SecurityManagerListener`) can be notified of failed PTU, and pass information about failed `service_type` to `ServiceStatusUpdateHandlerListener`

LuxoftAKutsan replied:
@mked-luxoft completely unclear interfaces. We need to rename them of join.

mked-luxoft replied:
@LuxoftAKutsan fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

LuxoftAKutsan commented on src/components/application_manager/include/application_manager/message_helper.h:
```diff
@@ -88,6 +88,19 @@ class MessageHelper {
+   * @param update_reason - enum value representing update_reason
+   * @return smart object containing on status update notification
+   */
+  static smart_objects::SmartObjectSPtr CreateOnServiceStatusUpdateNotification(
```
@mked-luxoft amount of parameters does not math description

mked-luxoft replied:
@LuxoftAKutsan  fixed in 007c20e6953023dc50553d14a4eaa354ff325fc0

LuxoftAKutsan commented on src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/on_service_status_update_notification.h:
```diff
@@ -0,0 +1,81 @@
+  /**
+   * @brief OnServiceStatusUpdateNotification class destructor
+   **/
+  virtual ~OnServiceStatusUpdateNotification();
```
@mked-luxoft OVERRIDE.

LuxoftAKutsan commented on src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/decrypt_certificate_response.cc:
```diff
@@ -60,6 +60,8 @@ void DecryptCertificateResponse::Run() {
 
   const bool is_succeeded = hmi_apis::Common_Result::SUCCESS == code;
 
+  smart_objects::SmartObjectSPtr notification;
```
@mked-luxoft is this variable used?

LuxoftAKutsan commented on src/components/application_manager/src/application_impl.cc:
```diff
@@ -507,13 +507,23 @@ void ApplicationImpl::StartStreaming(
       LOG4CXX_TRACE(logger_, "Video streaming not approved");
       MessageHelper::SendNaviStartStream(app_id(), application_manager_);
+      auto on_service_status_update_notification =
+          MessageHelper::CreateOnServiceStatusUpdateNotification(
```
@mked-luxoft and where do you specify actual status?

mked-luxoft replied:
@LuxoftAKutsan Since `OnServiceUpdate` requires only `service_type` as a mandatory parameter, it is the only value that is set in `MessageHelper::CreateOnServiceStatusUpdateNotification`. Other parameters are set outside this function if needed.

LuxoftAKutsan commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -138,6 +138,24 @@ namespace jhs = ns_smart_device_link::ns_json_handler::strings;
 using namespace ns_smart_device_link::ns_smart_objects;
 
+hmi_apis::Common_ServiceType::eType GetHMIServiceType(
+    protocol_handler::ServiceType service_type) {
```
@mked-luxoft I can see ServiceType is part of PH. PH depends from API, so I suppose that this funciton can be moved to PH

mked-luxoft replied:
@LuxoftAKutsan  fixed in 007c20e6953023dc50553d14a4eaa354ff325fc0. One note though. I don't think that moving declaration of this function to `protocol_handler.h` is a good idea, since we need to include `HMI_API.h` and as a result add include path to it for every translation unit that includes `protocol_handler.h`. Hence, i moved definition of this function to `service_status_update_handler.cc`

LuxoftAKutsan commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -3361,6 +3413,9 @@ void ApplicationManagerImpl::ProcessReconnection(
 void ApplicationManagerImpl::OnPTUFinished(const bool ptu_result) {
   LOG4CXX_AUTO_TRACE(logger_);
   if (!ptu_result) {
+    // auto notification =
```
@mked-luxoft remove commented

LuxoftAKutsan commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler.h:
```diff
@@ -0,0 +1,60 @@
+  void OnSuccessfulServiceUpdate(protocol_handler::ServiceType service_type,
+                                 const bool accepted);
+
+  void AddListener(ServiceStatusUpdateHandlerListener* listener);
```
@mked-luxoft is there should be multiple listeners?

mked-luxoft replied:
@LuxoftAKutsan the intent was to make container of listeners for future extensibility. If it's not relevant in the scope of this PR, i will remove this.

LuxoftAKutsan replied:
@mked-luxoft this is over engineering.  Please remove

mked-luxoft replied:
@LuxoftAKutsan fixed in 396e47180838b9bc07e5d435b11547170d25128d

LuxoftAKutsan commented on src/components/protocol_handler/src/service_status_update_handler.cc:
```diff
@@ -0,0 +1,79 @@
+    protocol_handler::ServiceType service_type) {
+  using namespace hmi_apis;
+  using namespace protocol_handler;
+  switch (service_type) {
```
@mked-luxoft please try to move it to policy handler

mked-luxoft replied:
@LuxoftAKutsan  it's been moved to service status update handler

BSolonenko commented on src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/on_service_status_update_notification.h:
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2018, Ford Motor Company
```
@mked-luxoft Copyright (c) **2019**, Ford Motor Company

BSolonenko commented on src/components/application_manager/src/message_helper/message_helper.cc:
```diff
@@ -1990,6 +1990,19 @@ smart_objects::SmartObjectSPtr MessageHelper::CreateNegativeResponse(
+
+  auto& msg_ref = (*on_status_update_notification);
+
+  msg_ref[strings::msg_params][hmi_notification::service_type] = service_type;
```
@mked-luxoft Why do you use msg_ref, but do not use on_status_update_notification?

BSolonenko commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler.h:
```diff
@@ -0,0 +1,60 @@
+/*
+ Copyright (c) 2013, Ford Motor Company
```
@mked-luxoft 2019

BSolonenko commented on src/components/protocol_handler/src/service_status_update_handler.cc:
```diff
@@ -0,0 +1,79 @@
+  listeners_.push_back(listener);
+}
+
+}  // namespace protocol_handler
```
@mked-luxoft please add empty line

LuxoftAKutsan commented on src/components/protocol_handler/src/service_status_update_handler.cc:
```diff
@@ -0,0 +1,79 @@
+  const auto service_update =
+      hmi_apis::Common_ServiceUpdateReason::INVALID_CERT;
+
+  for (const auto listener : listeners_) {
```
@mked-luxoft const auto& listener

mked-luxoft replied:
@LuxoftAKutsan fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

LuxoftAKutsan commented on src/components/security_manager/src/security_manager_impl.cc:
```diff
@@ -388,6 +388,18 @@ void SecurityManagerImpl::OnSystemTimeArrived(const time_t utc_time) {
+    LOG4CXX_ERROR(logger_, "listeners arrays IS EMPTY!");
+  }
+  std::list<SecurityManagerListener*>::iterator it = listeners_.begin();
+  while (it != listeners_.end()) {
```
@mked-luxoft why not foreach?

mked-luxoft replied:
@LuxoftAKutsan fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

LuxoftAKutsan reviewed (changes requested):
@mked-luxoft  could you please draw UL diagram of relations of ServiceStatusUpdateHandler, ProtocolHandler, SecurityManager,ApplicationManager and attach it to PR description.

AKalinich-Luxoft commented on src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/on_service_status_update_notification.h:
```diff
@@ -0,0 +1,81 @@
+  /**
+   * @brief OnServiceStatusUpdateNotification class constructor
+   **/
+  OnServiceStatusUpdateNotification(
```
@mked-luxoft please describe all other parameters

mked-luxoft replied:
@AKalinich-Luxoft  fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

AKalinich-Luxoft commented on src/components/application_manager/src/application_impl.cc:
```diff
@@ -507,13 +507,29 @@ void ApplicationImpl::StartStreaming(
     if (!video_streaming_approved()) {
       LOG4CXX_TRACE(logger_, "Video streaming not approved");
       MessageHelper::SendNaviStartStream(app_id(), application_manager_);
+      auto on_service_status_update_notification =
```
@mked-luxoft I would suggest you to create another helper function in MessageHelper which is accepting 2-3 parameters, and sends message created by `CreateOnServiceStatusUpdateNotification` with all required parameters. Something like:
```
SendOnServiceStatusUpdate(rpc_service_, AUDIO, REQUEST_RECEIVED)
```

mked-luxoft replied:
@AKalinich-Luxoft fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

mked-luxoft replied:
@AKalinich-Luxoft added two overloads in ad2041755a24692ab28fb7fe5506dbb685089057

AKalinich-Luxoft commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler.h:
```diff
@@ -0,0 +1,60 @@
+  void OnSuccessfulServiceUpdate(protocol_handler::ServiceType service_type,
+                                 const bool accepted);
+
+  void AddListener(ServiceStatusUpdateHandlerListener* listener);
```
@mked-luxoft If you are going to implement `AddListener` then it makes sense to add `RemoveListener` as well. 
Also suggestion to use shared pointers instead of raw pointers:
```
typedef std::vector<ServiceStatusUpdateListenerSPtr> ListenersList;
```

AKalinich-Luxoft commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler.h:
```diff
@@ -0,0 +1,60 @@
+      ServiceStatusUpdateListenerSPtr;
+  typedef std::vector<ServiceStatusUpdateHandlerListener*> ListenersList;
+
+  void OnPTUFailed(protocol_handler::ServiceType service_type);
```
@mked-luxoft please provide a description for all methods described in this interface. It is required for understanding how to use it

mked-luxoft replied:
@AKalinich-Luxoft fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

AKalinich-Luxoft commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler.h:
```diff
@@ -0,0 +1,60 @@
+
+  void OnPTUFailed(protocol_handler::ServiceType service_type);
+  void OnGetSystemTimeExpired(protocol_handler::ServiceType service_type);
+  void OnCertInvalid(protocol_handler::ServiceType service_type);
```
@mked-luxoft don't forget about `const` for input parameters. I believe it can be applied for all functions described here

AKalinich-Luxoft commented on src/components/application_manager/src/policies/policy_handler.cc:
```diff
@@ -1085,6 +1085,7 @@ bool PolicyHandler::ReceiveMessageFromSDK(const std::string& file,
         correlation_id, vehicle_data_args, application_manager_);
   } else {
     LOG4CXX_WARN(logger_, "Exchange wasn't successful");
+    application_manager_.ProcessFailedPTU();
```
@mked-luxoft I believe you can use `OnPTUFinished(ret)` to call `ProcessFailedPTU`/`ProcessSuccessfulStatusUpdate` in AM

mked-luxoft replied:
@AKalinich-Luxoft fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

AKalinich-Luxoft commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler_listener.h:
```diff
@@ -0,0 +1,58 @@
+
+class ServiceStatusUpdateHandlerListener {
+ public:
+  virtual void ProcessFailedStatusUpdate(
```
@mked-luxoft please provide description for these functions

mked-luxoft replied:
@AKalinich-Luxoft  fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

AKalinich-Luxoft commented on src/components/protocol_handler/src/handshake_handler.cc:
```diff
@@ -72,6 +75,9 @@ bool HandshakeHandler::GetPolicyCertificateData(std::string& data) const {
 void HandshakeHandler::OnCertificateUpdateRequired() {}
 
 bool HandshakeHandler::OnHandshakeFailed() {
+  service_status_update_handler_->OnGetSystemTimeExpired(
```
@mked-luxoft is that the only reason when `OnHandshakeFailed` called?

mked-luxoft replied:
@AKalinich-Luxoft  i think the only place where it is called is `BasicCommunicationGetSystemTimeRequest::onTimeOut`

AKalinich-Luxoft commented on src/components/protocol_handler/src/handshake_handler.cc:
```diff
@@ -102,6 +112,13 @@ bool HandshakeHandler::OnHandshakeDone(
+  if (!success) {
+    service_status_update_handler_->OnCertInvalid(context_.service_type_);
+  } else {
+    service_status_update_handler_->OnSuccessfulServiceUpdate(
```
@mked-luxoft please invert this statement. Successful result should be first

mked-luxoft replied:
@AKalinich-Luxoft fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

AKalinich-Luxoft commented on src/components/protocol_handler/src/service_status_update_handler.cc:
```diff
@@ -0,0 +1,79 @@
+  }
+}
+
+void ServiceStatusUpdateHandler::OnGetSystemTimeExpired(
```
@mked-luxoft I believe it is possible to create just one callback function with enum as a second parameter like:
```
enum class ServiceStatus {kPTUFailed, kPTUStarted, kPTUFinished, kInvalidCert, kInvalidTime}

OnServiceStatusUpdate(protocol_handler::ServiceType service_type, ServiceStatus status)
```

Depending on second parameter, you can easily determine `service_event` and `service_update` parameters. Try to reduce callbacks count as less as possible

mked-luxoft replied:
@AKalinich-Luxoft fixed in 4a605338d184abfca2cd4da602518727e29cb9e8

LuxoftAKutsan commented on src/components/application_manager/include/application_manager/application_manager_impl.h:
```diff
@@ -520,6 +520,7 @@ class ApplicationManagerImpl
    * @brief ProcessStatusUpdate is called on service status update received
    */
   void ProcessStatusUpdate(
+      const uint32_t connection_key,
```
@mked-luxoft function parameters does not match description

mked-luxoft replied:
@LuxoftAKutsan fixed in 396e47180838b9bc07e5d435b11547170d25128d

LuxoftAKutsan commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -1530,15 +1535,47 @@ void ApplicationManagerImpl::OnServiceEndedCallback(
   }
 }
 
+void AddAppIDForRPCService(bool& is_first_rpc_service_accepted,
```
@mked-luxoft output parameter was not marked as output, and it is in the begining.  Why nit just return it?

LuxoftAKutsan replied:
@mked-luxoft also function requires description

mked-luxoft replied:
@LuxoftAKutsan  changed to member function with more clear interface 396e47180838b9bc07e5d435b11547170d25128d

LuxoftAKutsan commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -1530,15 +1535,47 @@ void ApplicationManagerImpl::OnServiceEndedCallback(
+                           const uint32_t connection_key) {
+  if (!is_first_rpc_service_accepted &&
+      service_event == hmi_apis::Common_ServiceEvent::REQUEST_ACCEPTED) {
+    is_first_rpc_service_accepted = true;
```
@mked-luxoft use early return

mked-luxoft replied:
@LuxoftAKutsan fixed in 396e47180838b9bc07e5d435b11547170d25128d

LuxoftAKutsan commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -1530,15 +1535,47 @@ void ApplicationManagerImpl::OnServiceEndedCallback(
       service_event;
+
+  if (service_type == hmi_apis::Common_ServiceType::RPC) {
+    AddAppIDForRPCService(is_first_rpc_service_accepted,
```
@mked-luxoft is_first_rpc_service_accepted is output variable. No not see reason for it.

mked-luxoft replied:
@LuxoftAKutsan fixed in 396e47180838b9bc07e5d435b11547170d25128d

LuxoftAKutsan commented on src/components/protocol_handler/src/handshake_handler.cc:
```diff
@@ -115,10 +117,14 @@ bool HandshakeHandler::OnHandshakeDone(
-        context_.service_type_, ServiceStatus::CERT_INVALID);
+        this->connection_key(),
+        context_.service_type_,
+        ServiceStatus::CERT_INVALID);
```
@mked-luxoft all parameters except last one are the same. 
Propose : 
```cpp
const ServiceStatus service_status = success ? ServiceStatus::SERVICE_ACCEPTED ? ServiceStatus::CERT_INVALID;
service_status_update_handler_->OnServiceUpdate(this->connection_key(), context_.service_type_,service_status);
)
```

mked-luxoft replied:
@LuxoftAKutsan done in  396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/on_service_status_update_notification.h:
```diff
@@ -0,0 +1,85 @@
+  /**
+   * @brief Execute command
+   **/
+  virtual void Run();
```
@mked-luxoft Is this function inherited from base class?

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/on_service_status_update_notification.cc:
```diff
@@ -0,0 +1,66 @@
+}  // namespace hmi
+
+}  // namespace commands
+
```
@mked-luxoft Why did you add so many empty lines?

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -1418,6 +1418,10 @@ void ApplicationManagerImpl::OnServiceStartedCallback(
   std::vector<std::string> empty;
 
   if (type == kRpc) {
```
@mked-luxoft I would swap comparables here

mked-luxoft replied:
@AByzhynar  fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -1526,6 +1535,56 @@ void ApplicationManagerImpl::OnServiceEndedCallback(
+  (*notification)[strings::msg_params][hmi_notification::service_event] =
+      service_event;
+
+  if (service_type == hmi_apis::Common_ServiceType::RPC) {
```
@mked-luxoft swap comparables please

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/application_manager/src/policies/policy_handler.cc:
```diff
@@ -1592,6 +1592,11 @@ uint32_t PolicyHandler::TimeoutExchangeMSec() const {
+
+  std::for_each(listeners_.begin(),
+                listeners_.end(),
+                std::mem_fun(&PolicyHandlerObserver::OnPTUTimeoutExceeded));
```
@mked-luxoft `mem_fun` is deprecated . CAn we use `mem_fn` from C++11 ?

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler.h:
```diff
@@ -0,0 +1,96 @@
+#ifndef SRC_COMPONENTS_PROTOCOL_HANDLER_INCLUDE_PROTOCOL_HANDLER_SERVICE_STATUS_UPDATE_HANDLER_H_
+#define SRC_COMPONENTS_PROTOCOL_HANDLER_INCLUDE_PROTOCOL_HANDLER_SERVICE_STATUS_UPDATE_HANDLER_H_
+
+#include "service_status_update_handler_listener.h"
```
@mked-luxoft Add containing directory to path

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler.h:
```diff
@@ -0,0 +1,96 @@
+/**
+ *@brief  ServiceUpdateFailureReason helper enum containing reasons for service
+ *status to be updated
+  **/
```
@mked-luxoft Check indents

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler.h:
```diff
@@ -0,0 +1,96 @@
+  typedef std::list<ServiceStatusUpdateHandlerListener*> ListenersList;
+
+  /**
+  *@brief OnServiceUpdate callback that is invoked in case of
```
@mked-luxoft Wrong indentation

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler.h:
```diff
@@ -0,0 +1,96 @@
+
+}  // namespace protocol_handler
+
+#endif  // SRC_COMPONENTS_APPLICATION_MANAGER_INCLUDE_APPLICATION_MANAGER_SERVICE_STATUS_UPDATE_HANDLER_H_
```
@mked-luxoft 
Please add an empty line here

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler_listener.h:
```diff
@@ -0,0 +1,75 @@
+/*
```
@mked-luxoft Check license header correctness. Compare to other files headers

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/protocol_handler/include/protocol_handler/service_status_update_handler_listener.h:
```diff
@@ -0,0 +1,75 @@
+namespace protocol_handler {
+/**
+ *@brief Converts service type enum value from protocol_handler to hmi_apis.
+ *
```
@mked-luxoft 
Redundant empty line

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/protocol_handler/src/service_status_update_handler.cc:
```diff
@@ -0,0 +1,86 @@
+    case SERVICE_TYPE_NAVI: {
+      return Common_ServiceType::VIDEO;
+    }
+    default: { return Common_ServiceType::INVALID_ENUM; }
```
@mked-luxoft Divide to several lines as you did in other cases

mked-luxoft replied:
@AByzhynar although check_style seemed to alright with this, fixed in 396e47180838b9bc07e5d435b11547170d25128d

AByzhynar commented on src/components/security_manager/src/security_manager_impl.cc:
```diff
@@ -388,6 +388,16 @@ void SecurityManagerImpl::OnSystemTimeArrived(const time_t utc_time) {
+void SecurityManagerImpl::ProcessFailedPTU() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  if (listeners_.empty()) {
+    LOG4CXX_ERROR(logger_, "listeners arrays IS EMPTY!");
```
@mked-luxoft Don't we need to return in case of empty listeners?

mked-luxoft replied:
@AByzhynar fixed in 396e47180838b9bc07e5d435b11547170d25128d

LuxoftAKutsan commented on src/components/application_manager/include/application_manager/application_manager_impl.h:
```diff
@@ -1371,6 +1407,7 @@ class ApplicationManagerImpl
   bool is_vr_session_strated_;
   bool hmi_cooperating_;
   bool is_all_apps_allowed_;
+  bool is_first_rpc_service_accepted_;
```
@mked-luxoft is it applicable for app application in one variable?  you keep it in applicaiton Manager

mked-luxoft replied:
@LuxoftAKutsan as i understand first rpc service can be started before registering any application. that is why i placed it to app manager

LuxoftAKutsan commented on src/components/application_manager/include/application_manager/policies/regular/policy_handler_observer.h:
```diff
@@ -50,6 +50,8 @@ class PolicyHandlerObserver {
 
   virtual void OnPTUFinished(const bool ptu_result) {}
 
+  virtual void OnPTUTimeoutExceeded() {}
```
@mked-luxoft can we make it pure virtual?

mked-luxoft replied:
@LuxoftAKutsan yes, we can, but as i understand the intent behind this is not to place similar stubs in all the instances that implement this interface and dont need this particular function. this is why this function has empty body here.

LuxoftAKutsan commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -1526,6 +1538,51 @@ void ApplicationManagerImpl::OnServiceEndedCallback(
+      hmi_apis::Common_ServiceEvent::REQUEST_ACCEPTED == service_event) {
+    is_first_rpc_service_accepted_ = true;
+    return false;
+  } else {
```
@mked-luxoft if you use return, else condition is redundant

LuxoftAKutsan commented on src/components/application_manager/src/policies/policy_handler.cc:
```diff
@@ -1592,6 +1592,11 @@ uint32_t PolicyHandler::TimeoutExchangeMSec() const {
+
+  std::for_each(listeners_.begin(),
+                listeners_.end(),
+                std::mem_fn(&PolicyHandlerObserver::OnPTUTimeoutExceeded));
```
@mked-luxoft why do we need mem_fn ? why not just &PolicyHandlerObserver::OnPTUTimeoutExceeded ?

mked-luxoft replied:
@LuxoftAKutsan as far as i understand, pointers to member functions need to wrapped in `std::mem_fn` for `std::for_each` to work properly

LuxoftAKutsan commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -1526,6 +1538,51 @@ void ApplicationManagerImpl::OnServiceEndedCallback(
+  (*notification)[strings::msg_params][hmi_notification::service_event] =
+      service_event;
+
+  if (ShouldAddAppIDForService(service_type, service_event)) {
```
@mked-luxoft looks like you adding app_id optionally according to following part of the proposal 
```
Note that the only time when SDL would not be able provide the appID would be during the first StartService request for the RPC service before RAI was sent.
```
The point there is that SDL is **unable** to provide app_id in case applicaiton start secure service before RAI.

So in the code in that place you probably have no ability to get the connection_key. Does it mean that `connection_key` arg is broken? or NULL? I would better make this argument optional in the ProcessServiceStatusUpdate function, and find out place where you can't pass it.

mked-luxoft replied:
@LuxoftAKutsan as far as i understand, mobile app_id doubles also as a connection key. in turn, it is obtained by connection handler from message packet, and passed to session context, which is in turn passed to protocol handler. as long as data in the message packet is valid, connection id will not be broken. then, `CommandImpl::ReplaceMobileWithHMIAppId` is used to convert app id for HMI notification. if it fails, command will not be run. Moreover, this function should restrict adding app id to notification for cases where app id can be invalid

vkushnirenko-luxoft commented:
Rebuild required

LuxoftAKutsan reviewed (approved):
Please take care about MessageHelper::CreateOnServiceStatusUpdateNotification and first_rpc_service_accepted redundant flag in the nearest bug fix

mked-luxoft committed:
Added OnServiceUpdate notification

mked-luxoft committed:
Added handling of service status update

- introduced ServiceStatusUpdateHandler and ServiceStatusUpdateHandlerListener interfaces
- added lacking interface functions to corresponding entities
- added notification of listeners
